### PR TITLE
feat(new source): add AWS SQS source

### PR DESCRIPTION
**Pull request by fuchsnj**
closes https://github.com/vectordotdev/vector/issues/6830
This introduces a new AWS SQS source.

This also is the first source to start using the AWS official rust SDK. There are a few small rough edges around config between the new/old SDK, but for the most part it just works. The biggest downside I've seen is that the new SDK currently doesn't allow using a custom file for auth (it can only use the default file locations).

E2E acking, and setting concurrency limits were included. Emitting latency metrics was considered out of scope for this PR. This is something we are looking at introducing in a more generic way in the future.

`EventsReceived` was also moved to `vector_core` (we are slowly moving all of the "core" events into `vector_core`)

### CODE REVIEW

File: src/aws/rusoto/region.rs [hunk 1]
```diff
@@ -1,32 +1,12 @@
...
-    }
-
-    pub const fn with_endpoint(endpoint: String) -> Self {
-        Self {
-            region: None,
-            endpoint: Some(endpoint),
-        }
-    }
-}
+//TODO: backwards compat - use the type directly instead of from here
```
Comment (blt) [inline src/aws/rusoto/region.rs:8]: Intentionally left TODO?

File: src/sources/aws_sqs/integration_tests.rs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+use std::str::FromStr;
+use std::time::Duration;
+use tokio::time::timeout;
+
+// fn gen_queue_url(name: &str) -> String {
+//     format!(
+//         "https://sqs.us-east-1.amazonaws.com/123456789012/test-{}",
+//         name
+//     )
+// }
```
Comment (blt) [inline src/sources/aws_sqs/integration_tests.rs:27]: Left commented code here.

File: website/cue/reference/components/sources/aws_sqs.cue [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+//				enabled_default:        false
+//			}
+//			from: components._kafka.features.collect.from
+//		}
+//		multiline: enabled: false
+//		codecs: {
+//			enabled:         true
+//			default_framing: "bytes"
+//		}
+//	}
```
Comment (blt) [inline website/cue/reference/components/sources/aws_sqs.cue:23]: Intentionally left?

**Review by blt — COMMENTED** [review]
This seems straightforward and uncontroversial to me. I've got some comments but nothing major.

File: website/cue/reference/components/sources/aws_sqs.cue (hunk 3 above)
Comment (fuchsnj) [reply website/cue/reference/components/sources/aws_sqs.cue:23]: This is what happens when I write most of the docs on a Friday, and finish on a Monday :)
Need to figure out what all these features mean and use the correct ones...

**Review by blt — APPROVED** [review]

File: website/cue/reference/components/sources/aws_sqs.cue [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+		//  client_concurrency: {
+		//   common: true
+		//   description: "How many clients are receiving / acking SQS messages. Increasing may allow higher throughput. Note: the default is 1 / CPU core"
+		//   required: false
+		//   warnings: []
+		//   type: uint: {
+		//    default: 1
+		//    unit: "# of clients"
+		//   }
+		//  }
```
Comment (jszwedko) [inline website/cue/reference/components/sources/aws_sqs.cue:74]: Did we end up leaving this option out?

File: src/aws/aws_sdk/auth.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            } => SharedCredentialsProvider::new(Credentials::from_keys(
+                access_key_id,
+                secret_access_key,
+                None,
+            )),
+            AwsAuthentication::File {
+                credentials_file,
+                profile,
+            } => {
+                warn!("Overriding the credentials file is not supported. `~/.aws/config` and `~/.aws/credentials` will be used instead of \"{}\"", credentials_file);
```
Comment (jszwedko) [inline src/aws/aws_sdk/auth.rs:27]: Did we note this in the docs for this source?

**Comment by jszwedko** [comment]
Just noticing that I don't think we added this new source's integration tests to the integration test workflow: https://github.com/vectordotdev/vector/blob/master/.github/workflows/integration-test.yml

I wonder if we could get the makefile to output the integration test targets and dynamically build the set to run (similar to the soak tests dynamically building their matrix) 🤔

File: src/aws/aws_sdk/auth.rs (hunk 5 above)
Comment (fuchsnj) [reply src/aws/aws_sdk/auth.rs:27]: I wasn't sure of a good place to put this. All the AWS docs are shared, and I think this is a pretty minor edge case that very few would hit.

File: website/cue/reference/components/sources/aws_sqs.cue (hunk 4 above)
Comment (fuchsnj) [reply website/cue/reference/components/sources/aws_sqs.cue:74]: this was not intentional. I'll add this back in.

**Comment by fuchsnj** [comment]
> Just noticing that I don't think we added this new source's integration tests to the integration test workflow: https://github.com/vectordotdev/vector/blob/master/.github/workflows/integration-test.yml
> 
> I wonder if we could get the makefile to output the integration test targets and dynamically build the set to run (similar to the soak tests dynamically building their matrix) thinking

`aws` covers _all_ of the aws integration tests (including this one).

**Comment by jszwedko** [comment]
Aha, you are right! Thanks!

**Pull request merged**